language/bytecode_verifier/src/unused_entries.rs

```diff
@@ -0,0 +1,114 @@
+use types::vm_error::{StatusCode, VMStatus};
+use vm::{
+    access::ModuleAccess,
+    errors::verification_error,
+    file_format::{Bytecode, CompiledModule, StructFieldInformation},
+    IndexKind,
+};
+
+pub struct UnusedEntryChecker<'a> {
+    module: &'a CompiledModule,
+
+    field_defs: Vec<bool>,
+    locals_signatures: Vec<bool>,
+    type_signatures: Vec<bool>,
+}
+
+impl<'a> UnusedEntryChecker<'a> {
+    pub fn new(module: &'a CompiledModule) -> Self {
+        Self {
+            module,
+            field_defs: vec![false; module.field_defs().len()],
+            locals_signatures: vec![false; module.locals_signatures().len()],
+            type_signatures: vec![false; module.type_signatures().len()],
+        }
+    }
+
+    fn traverse_function_defs(&mut self) {
+        use Bytecode::*;
+
+        for func_def in self.module.function_defs() {
+            if func_def.is_native() {
+                continue;
+            }
+            self.locals_signatures[func_def.code.locals.0 as usize] = true;
+
+            for bytecode in &func_def.code.code {
+                match bytecode {
+                    Call(_, idx)
+                    | Pack(_, idx)
+                    | Unpack(_, idx)
+                    | MutBorrowGlobal(_, idx)
+                    | ImmBorrowGlobal(_, idx)
+                    | Exists(_, idx)
+                    | MoveToSender(_, idx)
+                    | MoveFrom(_, idx) => {
+                        self.locals_signatures[idx.0 as usize] = true;
+                    }
+                    _ => (),
+                }
+            }
+        }
+    }
+
+    fn traverse_struct_defs(&mut self) {
```

Comment:
I understand the desire to have all unused entry checking in one place.  But, in this case, we are trading off a less efficient solution for that desire.  I am not convinced if this is indeed better.